_posts/outbox-what-and-why.md

```diff
@@ -0,0 +1,155 @@
+---
+layout: post
+title: "What's an Outbox and why do we need it? Hint: it's about data integrity"
+author: Mauro Servienti
+synopsis: "Distributed systems are ugly beasts sometimes. They hide subtle tricks that can lead to data loss and system corruption. The Outbox pattern helps address a couple of them."
+enable_mermaid: true
+header_image: /img/posts/outbox-what-and-why/header.jpg
+tags:
+- messaging
+- distributed-systems
+---
+
+When the software industry invented transactions and distributed transactions, the game changed. Suddenly, it was possible to guarantee that a set of operations was atomic. An all-or-none approach became the norm, greatly simplifying the engineers' work. Doing backflips to get consistency became a thing of the past. The transaction coordinator took responsibility for all that.
+
+How does that influence a message-based system? Let's take a look at the following diagram:
+
+<div class="mermaid">
+graph LR
+    E[Endpoint] -- store data --> DB[(database)]
+    E -- send message --> MSMQ
+</div>
+
+The presented endpoint needs to store some data and send a message. It uses MSMQ and SQL Server. At runtime, they will both enlist in a distributed transaction, making it so that both operations will be atomic and consistent. If for some reason, one of the operations fails, everything is rolled back. The system won't send messages and won't change any data. That's very handy, and it makes things so simple.
+
+It comes at a cost, though. Transactions are expensive, and distributed transactions are even more costly. They have a performance impact because getting all the participants into an agreement takes time. As you can imagine, the more participants, the greater the effect. Last but not least, transaction protocols rely on high network stability. Poor network connectivity hinders participants' ability to participate in the distributed transaction coordination process. That means it's unlikely cloud environments support distributed transaction protocols.
+
+All that means the above-shown diagram only works in a cloud environment if the database server is local to the same machine where the MSMQ is. This sounds illogical because that means the only option to use distributed transactions is to have all participants on the same box, which is precisely what we don't want.
```

Comment:
```suggestion
All this means that the above diagram only works in a cloud environment if the database server is local to the same machine where the MSMQ machine is. This sounds illogical because that means the only option to use distributed transactions is to have all participants on the same box, which is precisely what we don't want in the cloud.
```

Reply:
```suggestion
All this means that the above diagram only works in a cloud environment if the database server runs on the same machine where the MSMQ is installed. This sounds illogical because that means the only option to use distributed transactions is to have all participants on the same box, which is precisely what we don't want in the cloud.
```